Deprecation

silene opened the pull request:
These patches remove support for versions of Coq up to 8.4, they remove deprecated compatibility options, and they deprecate compatibility options for versions of Coq up to 8.5. The gain on the code base will only be noticeable in 8.8, once all the related code can finally be removed, but deprecation has to happen in 8.7.

herbelin commented on engine/namegen.ml:
```diff
@@ -404,14 +404,13 @@ let rename_bound_vars_as_displayed avoid env c =
 let h_based_elimination_names = ref false
 
-let use_h_based_elimination_names () =
-  !h_based_elimination_names && Flags.version_strictly_greater Flags.V8_4
```
Actually, this option was never activated by default because I thought it would have led to too much incompatibilities. Moreover, we can do even better in naming hypotheses produced by destruct/induction. So, not only I see this option as only a step towards a big picture, but I don't think it is reasonable to change the naming model until we have a way to globally change the tactic language using commands to invent such as "Require Tactics (version 8.5)" or similar.

(Note: the policy that some people advocate of naming everything by hand is also fine to me!)

In particular, I would not deprecate it.

There is another option in this situation, it is the "Structural Injection" which is off by default.

silene replied:
Note that I am not suggesting to change the option value in any way. If the option has always been disabled and if it ever gets removed, then it is the disabled value that would be used, obviously. So the naming model would certainly not change due to deprecation.

JasonGross commented:
I'd like to express a preference for keeping 8.4 compatibility around until either (a) we freeze a version of Ltac 1.0 as a plugin (@ppedrot, any updates on when this is happening?), or (b) the version of Coq in Ubuntu LTS is newer than 8.4 (this would be April 2018).

It's still unclear to me what the expected compatibility story is.  With this change, it seems to be "expect to spend a week rewriting all of your tactic automation once every two releases; compatibility options are only for helping the transition happen in parts during that week"?

maximedenes commented:
@silene So what should we do with this PR? Is more discussion needed? I remember we decided during a working group to support compatibility options for two releases.

@JasonGross I understand that the story for compatibility is not good. But I don't see how you could expect a nice story as soon as your code fundamentally relies on undocumented behavior and undefined semantics.

One thing we should do at some point is have a look at the remaining 8.4 code from MIT and understand if we can help with porting it. It is a service the Consortium could provide.

JasonGross commented:
@maximedenes I'd prefer to not rely on undocumented behavior or undefined semantics, but I have trouble figuring out other ways that work well.  (The latest example of this is that I needed to unfold a constant in 8.5 that didn't exist in 8.4, in code that needs to run in both 8.4 and 8.5; I took advantage of the fact that `Import` refuses to mask absolute names, so, e.g., `Coq.Init.Logic.eq` will always refer to the one in the standard library, and never to `Top.Coq.Init.Logic.eq`; is this undocumented behavior?)  In any case, is there any hope of an LtacLint, that will tell me when my tactics rely on autogenerated names, undefined behavior, or undocumented semantics?

> One thing we should do at some point is have a look at the remaining 8.4 code from MIT and understand if we can help with porting it. It is a service the Consortium could provide.

That would be great!  Current issues are adapting the plugin to not drop universes (I mentioned this in an email), and `cbv` running out of memory.  Is there any way I could get a debug mode for `cbv`?  For example, it would be very useful even to just have a live printout of every identifier that's encountered which isn't unfolded (better yet to have a live printout that only prints an identifier if it hasn't been encountered before).

silene commented:
> So what should we do with this PR? @maximedenes 

As far as I am concerned, this PR is still fine. I don't agree with @JasonGross that Coq 8.7 should be compatible with Coq 8.4, and, as you said, neither did the working group.

For "Structural Injection", I don't mind undeprecating it. For "Standard Proposition Elimination Names", I would still keep it deprecated, if only because it has never been documented. If it gets documented by the time of Coq 8.8, then we can undeprecate it.

maximedenes commented:
I feel I'm missing something. Does this PR touch the stdlib? If not, how could it ever compile since the stdlib uses 8.3 compat notations?

silene commented:
> If not, how could it ever compile since the stdlib uses 8.3 compat notations?

Good point, I might have forgotten to push something. I can't check right now, since the corresponding local branch is not on my current computer. I will sort it out by next week (and I will also rebase it since it seems there have been some changes to the code).

maximedenes commented:
No worries. I had tried to remove those compat notations from stdlib at some point, but some of the renamings that it induces look very painful to me. So maybe it would be an opportunity to discuss that.

For instance, renaming `le_n_Sn` into `Nat.le_succ_diag_r` doesn't seem to go in the right direction to me.

By the way, another problem is that using these compatibility notations doesn't trigger any warning, I don't know why.

silene commented:
> but some of the renamings that it induces look very painful to me.

Don't expect too much from my patch. I seem to remember that I did no more than just bumping the version number on the notations.

herbelin commented:
Actually, it is a long time that I wonder why we want lemma names to be unique. After all, in real life theorems have multiple names, like Feit-Thompson and odd order theorem, like fundamental theorem of algebra or Alembert-Gauss theorem.

With modules, we also already have synonyms.

I don't see what could be wrong at a coexistence of multiple names for the same object. After all, there are different rules and different usages to build theorem names. E.g. `le_n_Sn` which nicely abbreviates the statement, while a name like `le_succ_diag_r` tries to say something more of the properties. Or, e.g `mul_0_l`, `mul_neutral`, `times_0_l`, `mult_0_l` which all are legitimate from a naming policy point of view. In particular, when doing a proof and  looking for a theorem name, it is common to try different possible pattern hoping that the theorem will have a name built on one of these patterns. If we accept different pattern policies in parallel, this could make theorem name finding faster at the end.

silene commented:
I have rebased the branch. I was not too happy with my original idea of modifying the `.v` files, so I have instead introduced a new constructor characterizing older versions. (This does not preclude revisiting the choices made for 8.3.) I have also reverted the deprecation of "Structural Injection" since the option was documented.

> If we accept different pattern policies in parallel, this could make theorem name finding faster at the end.

There is the issue of all the `Search` commands. I am not sure how to strike a good balance between the user getting results satisfying the current naming policy and the user not being flooded with results for all the naming policies.

herbelin commented:
> There is the issue of all the Search commands. I am not sure how to strike a good balance between the user getting results satisfying the current naming policy and the user not being flooded with results for all the naming policies.

Isn't this just a presentational issue? `Search` does not have to return twice Feit-Thompson or FTA or `mul_0_l` just because it has several names. I guess that having several names can be seen as an attribute of a definition. I guess that naming policies can also be seen as preferred parameters for a search. There is something smart to do with [theorem names](http://link.springer.com/chapter/10.1007%2F978-3-319-43144-4_28).

maximedenes commented:
I don't understand the argument in favor of different naming policies. If the policy is documented, the user knows how to look for something.

silene commented:
> Isn't this just a presentational issue?

Sure. But until it is solved, having concurrent naming policies will make the life of the user worse, not easier.

herbelin commented:
I'm confused, Search does not look for notations, so, we are anyway talking about a model to implement in some future...

Zimmi48 commented:
Everybody is citing this "What's in a theorem name" paper but I've read it recently and did not find it very enlightening. It's just a machine learning paper.
But I agree there is some thinking to do about theorem names, but my view would rather be that there should be the possibility to give "long names" to theorems (e.g. "Mean Value Theorem") and there could be several long names for a single theorem. I'm much more reluctant to have several short names though. As Maxime said, there should be a naming policy and it should be documented.
And `Search` is far from ideal in its current state. One problem e.g. is that it shows terms and their types but not if they are `Definition`, `Lemma`, `Theorem` or `Fact`. When using math-comp and searching for a theorem, you might find one that looks suitable but which was designed for internal use only. These "internal use only" theorems are marked as `Fact` but since `Search` does not report this information, it doesn't help much. We should collect somewhere a TODO list for `Search`. (Maybe I'll create a wiki page. I've also said I'd create such a wiki page for `typeclasses eauto`.)

ppedrot commented:
@Zimmi48 Authors of a library should use the Local modifier for such lemmas.

silene commented:
Just so that people do not have to look at the Travis log, the failures are jobs that run Coq with option `-compat 8.4`. I am not even sure why we ever decided it was a good idea to test on trunk the compatibility with 8.4.

maximedenes commented:
Summary of WG discussion: we ask @JasonGross if there is a plan to get rid of the `-compat 8.4` flag for bedrock and fiat, we merge this PR and move developments still relying on those flags to allowed failures.

maximedenes commented:
Sequencing: because this PR currently breaks development, I'll merge it after the other ones to give a chance to fix what can be.

JasonGross commented:
As I understand it, @achlipala plans to rewrite the still-relevant parts of Bedrock from scratch over the next academic year or so, so there is no plan to port the current Bedrock to not use `-compat 8.4`.

Fiat-parsers is already incompatible with 8.4, and I'm currently discussing with Adam what we want to do about the fact that, currently, part of fiat builds only with 8.4, and another part doesn't build with 8.4.  Regardless, I would be interested in getting fiat-parsers off of -compat 8.4.  Certainly, if you have a patch that takes fiat-crypto off -compat 8.4, I'll find a way to merge it.  I can also look into doing that myself; when are you looking to have this merged by?

Zimmi48 commented:
@JasonGross The freeze is June 1st. Given @maximedenes's comment (I wasn't there during this part of the WG), I would expect this to be the last PR merged before the freeze.

maximedenes commented:
@silene Could you rebase it? Or would you like me to do it?

silene commented:
Done. Unfortunately the testsuite does not pass because new "8.4 only" files have been added to trunk. In particular, I am not quite sure what to do about 7b56ef8.

JasonGross commented:
> In particular, I am not quite sure what to do about 7b56ef8.

It seems to be testing quoting.  Just bump it to compat 8.5, and it should work

maximedenes commented:
It would be better to pick an option that will not break again in the next version. I'll change it to `-w default`.
